src/Servers/Kestrel/perf/Microbenchmarks/Http3/QPackDecoderBenchmark.cs

```diff
@@ -0,0 +1,120 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Net.Http.QPack;
+using BenchmarkDotNet.Attributes;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http3;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Microbenchmarks;
+
+public class QPackDecoderBenchmark
+{
+    private static readonly byte[] _headerFieldLine_LargeLiteralValue;
+    private static readonly byte[] _headerFieldLine_LargeLiteralValue_Multiple;
+    private static readonly byte[] _headerFieldLine_Static_Multiple;
+
+    static QPackDecoderBenchmark()
+    {
+        string string8193 = new string('a', 8193);
+
+        var headers = new HttpResponseHeaders();
+        var enumerator = new Http3HeadersEnumerator();
+
+        AddLargeHeader(headers, 'a');
+        enumerator.Initialize(headers);
+        _headerFieldLine_LargeLiteralValue = GenerateHeaderBytes(enumerator);
+        headers.Reset();
+
+        AddLargeHeader(headers, 'a');
+        AddLargeHeader(headers, 'b');
+        AddLargeHeader(headers, 'c');
+        AddLargeHeader(headers, 'd');
+        AddLargeHeader(headers, 'e');
+        enumerator.Initialize(headers);
+        _headerFieldLine_LargeLiteralValue_Multiple = GenerateHeaderBytes(enumerator);
+        headers.Reset();
+
+        ((IHeaderDictionary)headers).ContentLength = 0;
+        ((IHeaderDictionary)headers).ContentType = "application/json";
+        ((IHeaderDictionary)headers).Age = "0";
+        ((IHeaderDictionary)headers).AcceptRanges = "bytes";
+        ((IHeaderDictionary)headers).AccessControlAllowOrigin = "*";
+        enumerator.Initialize(headers);
+        _headerFieldLine_Static_Multiple = GenerateHeaderBytes(enumerator);
+
+        static void AddLargeHeader(HttpResponseHeaders headers, char c)
+        {
+            var s = new string(c, 8193);
+            var success = headers.TryAdd(s, s);
+            if (!success)
+            {
+                throw new InvalidOperationException();
+            }
+        }
+
+        static byte[] GenerateHeaderBytes(Http3HeadersEnumerator enumerator)
+        {
+            var buffer = new byte[1024 * 1024];
+            var totalHeaderSize = 0;
+            var success = QPackHeaderWriter.BeginEncodeHeaders(enumerator, buffer, ref totalHeaderSize, out var length);
+            if (!success)
+            {
+                throw new InvalidOperationException();
+            }
+
+            return buffer[0..length].ToArray();
```

Comment:
Is the `.ToArray()` necessary to free up unused buffer space or something? I thought slicing created a copy.

Reply:
```suggestion
            return buffer[..length];
```
is enough here, and translated to [`RuntimeHelpers.GetSubArray`](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+ABATARgLABQGADAAQY4B0AMgJYB2AjgNyGEYDM5WpAwqQG9CpEeS7AAngBcYAbQC6pALLEAFJJkLSwAK4AzPTCgAaUgymkANjHoBzKQAsAlKQC8APm37DUWcUqU1naO8pQAKhAAglBQAIYSqk6sBKJi2tJyiko46hlaugZGpuZWNvbObp4FPn4BQeXyyamc6ZpZWLltXoUmZvQW9Y4uHt01dWUhyQC+QA=).